### PR TITLE
MainMenu functions (Demo playback related)

### PR DESCRIPTION
**Pull request by d3nwah**
tDemoSelect_4D0E10()
tLoadDemo_4D1040()
sub_FileName_4D1430()
sub_convert_iso_FileName_4D1660()
OptionsMenuBtnListener_4D1AB0

### CODE REVIEW

File: Source/AliveLib/stdlib.hpp [hunk 1]
```diff
@@ -25,3 +25,12 @@ inline T* alive_new(Args&&... args)
     }
     return nullptr;
 }
+
+inline int access_impl(char const* fileName, int accessMode)
```
Comment (paulsapps) [inline Source/AliveLib/stdlib.hpp]: move to cpp + add prototype instead of inlining this

File: Source/AliveLib/MainMenu.hpp [hunk 2]
```diff
@@ -282,7 +283,7 @@ class MainMenuController : public BaseAnimatedWithPhysicsGameObject
     __int16 field_228_res_idx;
     __int16 field_22A_anim_frame_num;
     int field_22C_T80;
-    __int16 field_230_fmv_level_index;
+    __int16 field_230_selected_demo_index;
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.hpp]: This is used for FMV and Demos at least from what I can see, I'd rename to demo_or_fmv or just selected_index or something like that

File: Source/AliveLib/MainMenu.hpp [hunk 3]
```diff
@@ -200,18 +200,19 @@ class MainMenuController : public BaseAnimatedWithPhysicsGameObject
...
-    
-    
-    EXPORT signed int tLoadGame_Input_4D3EF0(DWORD input);
+    
+    EXPORT signed int tLoadGame_Input_4D3EF0(DWORD input);
     EXPORT void tLoadGame_Render_4D44D0( int **pOt);
-    
+    
     EXPORT void tLoadGame_Load_4D42F0();
+    EXPORT static char CC sub_FileName_4D1430(char *a1);
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.hpp]: function name and argument name sucks

File: Source/AliveLib/MainMenu.hpp [hunk 4]
```diff
@@ -200,18 +200,19 @@ class MainMenuController : public BaseAnimatedWithPhysicsGameObject
...
+    EXPORT static char CC sub_convert_iso_FileName_4D1660(char *a1, char *a2);
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.hpp]: ditto, should be remove iso_99960_suffix or something

File: Source/AliveLib/MainMenu.cpp [hunk 5]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
 
+ALIVE_VAR(1, 0x561538, int, word_561538, 0);
+
+signed int MainMenuController::tDemoSelect_4D0E10(DWORD input)
+{
+    word_5C1BA0 = 0;
+    word_5C1B9C = 0;
+
+    int input_or_field_204 = input;
+    if (field_204_prev_pressed && field_204_prev_pressed == static_cast<int>(sInputObject_5BD4E0.field_0_pads[sCurrentControllerIndex_5C1BBE].field_0_pressed))
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: why is this casted?

File: Source/AliveLib/MainMenu.cpp [hunk 6]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+            if (byte_5C1BA2 > maxDemoId)
+            {
+                byte_5C1BA2 = 1;
+            }
+            demoId = byte_5C1BA2;
+        }
+        char file[32];
+        sprintf(file, "ATTR%04d.SAV", sDemos_5617F0[demoId].field_A_id);
+        ResourceManager::LoadResourceFile_49C170(file, 0);
+        BYTE **resource = ResourceManager::GetLoadedResource_49C2A0(1349810254, 0, 1u, 0);
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: use the resource manager constant for 1349810254

File: Source/AliveLib/MainMenu.cpp [hunk 7]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+        if (word_5C1B9C <= 0)
+        {
+            byte_5C1BA2++;
+            if (byte_5C1BA2 > maxDemoId)
+            {
+                byte_5C1BA2 = 1;
+            }
+            demoId = byte_5C1BA2;
+        }
+        char file[32];
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: = {}

File: Source/AliveLib/MainMenu.cpp [hunk 8]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+        sprintf(file, "ATTR%04d.SAV", sDemos_5617F0[demoId].field_A_id);
+        ResourceManager::LoadResourceFile_49C170(file, 0);
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: last argument is a pointer AFAIK, use nullptr

File: Source/AliveLib/MainMenu.cpp [hunk 9]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+            }
+        }
 
+        if (field_F4_resources.field_0_resources[0] == 0)
+        {
+            pResourceManager_5C1BB0->LoadingLoop_465590(0);
+        }
+        field_20_animation.Set_Animation_Data_409C80(247808, field_F4_resources.field_0_resources[1]);
+        ResourceManager::FreeResource_49C330(field_F4_resources.field_0_resources[0]);
+        field_F4_resources.field_0_resources[0] = 0;
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: nullptr

File: Source/AliveLib/MainMenu.cpp [hunk 10]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+        if (word_5C1B9C == 0)
+        {
+            demoId = byte_5C1BA2;
+        }
+        if (demoId >= maxDemoId)
+        {
+            demoId = 0;
+        }
+        int levelId = static_cast<int>(sDemos_5617F0[demoId].field_4_level);
+        char lvFilename[256];
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: = {}

File: Source/AliveLib/MainMenu.cpp [hunk 11]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+        strcpy(lvFilename, "ATTRACT");
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: this buffer looks like 2 local arrays? one with ATTRACT in and one with the lvl name ??

File: Source/AliveLib/MainMenu.cpp [hunk 12]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+};
+
+ALIVE_VAR(1, 0x5C1B9C, short, word_5C1B9C, 0);
+ALIVE_VAR(1, 0x5C1BA2, BYTE, byte_5C1BA2, 0);
+ALIVE_VAR(1, 0x55C128, int, dword_55C128, 0);
+
+signed int MainMenuController::tLoadDemo_4D1040(DWORD)
 {
-    NOT_IMPLEMENTED();
+    const int maxDemoId = 23;
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: where does 23 come from? Can it be an ALIVE_COUNTOF on an array instead?

File: Source/AliveLib/MainMenu.cpp [hunk 13]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+    }
+    else
+    {
+        return 0;
+    }
 }
 
-signed int MainMenuController::tsub_4D0E10(DWORD /*input*/)
+const __int16 word_5617F4[4] = {
+    1, 8, 5, 0
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: can this have a better name?

File: Source/AliveLib/MainMenu.cpp [hunk 14]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
 
-signed int MainMenuController::tsub_LoadSave_Input_4D1040(DWORD /*input*/)
+char MainMenuController::sub_FileName_4D1430(char* input)
 {
-    NOT_IMPLEMENTED();
+    if (input[0] == '\\')
+    {
+        input++;
+    }
+    char buffer[256];
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: = {}

File: Source/AliveLib/MainMenu.cpp [hunk 15]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+    MainMenuController::sub_convert_iso_FileName_4D1660(buffer, input);
 
-    return 0;
+    char fileName[256];
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: = {}

File: Source/AliveLib/MainMenu.cpp [hunk 16]
```diff
@@ -1672,21 +1672,271 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
     pResourceManager_5C1BB0->LoadingLoop_465590(FALSE);
 }
 
+char MainMenuController::sub_convert_iso_FileName_4D1660(char *a1, char *a2)
+{
+    char result = *a2;
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: this function is quite messy its probably easier to rewrite it. I think its just copying the input to the output up to the first found ; ?

File: Source/AliveLib/MainMenu.cpp (hunk 5 above)
Comment (d3nwah) [reply Source/AliveLib/MainMenu.cpp]: signed/unsigned mismatch

Comment (paulsapps) [reply Source/AliveLib/MainMenu.cpp]: field_0_pressed its a bit field so it must be unsigned, therefore if field_204_prev_pressed is signed it should probably be changed to unsigned.

File: Source/AliveLib/MainMenu.cpp [hunk 17]
```diff
@@ -1672,21 +1672,268 @@ void MainMenuController::tGame_BackStory_Or_NewGame_Unload_4D1BE0()
...
+        field_202_input_hold_down_timer = 15;
+        field_204_prev_pressed = sInputObject_5BD4E0.field_0_pads[sCurrentControllerIndex_5C1BBE].field_0_pressed;
+    }
+
+    if (input_or_field_204 & 5)
+    {
+        if (field_230_selected_entry_index > 0 && !field_254)
+        {
+            field_230_selected_entry_index--;
+            SFX_Play_46FBA0(0x34u, 35, 400, FP_FromInteger(1));
```
Comment (paulsapps) [inline Source/AliveLib/MainMenu.cpp]: The last argument to SFX_Play_46FBA0 is defaulted to 1 AFAIK, so you probably don't need all these.

File: Source/AliveLib/MainMenu.cpp (hunk 11 above)
Comment (d3nwah) [reply Source/AliveLib/MainMenu.cpp]: One is a part of the other, resolved by adding lvFilenameNoPrefix

**Pull request merged**